src/v/io/pager.cc

```diff
@@ -0,0 +1,231 @@
+     * page exists and has not yet been evicted. in all other cases, allocate a
+     * new page and return (page, false).
+     */
+    auto find_or_alloc_page = [this, offset] {
+        if (auto it = pages_.find(offset); it != pages_.end()) {
+            auto page = *it;
+            if (!page->data().empty()) {
+                return std::make_pair(std::move(page), true);
+            }
+            /*
+             * reusing the page is technically possible, but isn't yet done,
+             * in favor of the simplicity of re-creating the page with a
+             * specific known states. the cache hook is transferred so that we
+             * do not lose the cache statistics.
+             */
+            pages_.erase(it);
+            return std::make_pair(
+              alloc_page(page->offset(), page->cache_hook), false);
+        }
+        return std::make_pair(
+          alloc_page(seastar::align_down(offset, page_size)), false);
+    };
+
+    while (true) {
+        auto [page, found] = find_or_alloc_page();
+        if (found) {
+            /*
+             * the page may be in the cache index, but faulting. in this case we
+             * can wait until the fault is finished and try again.
+             */
+            if (page->test_flag(page::flags::faulting)) {
+                page::waiter waiter;
+                page->add_waiter(waiter);
+                co_await waiter.ready.get_future();
+                continue;
+            }
+            vassert(
+              page->size() > 0, "Empty page at offset {}", page->offset());
+            co_return page;
+        }
+
+        /*
+         * the page was not found in the index, so we need to fault it in.
+         *
+         * NOTE: due to a not yet known issue, when waiter is allocated on the
+         * stack, co_await on the waiter triggers clang-tidy check
+         * StackAddressEscape. this seems like a false positive, so allocate on
+         * the heap for now to trick it.
+         */
+        auto waiter = std::make_unique<page::waiter>();
+        page->set_flag(page::flags::faulting);
+        page->add_waiter(*waiter);
+
+        cache_->insert(*page);
+
+        auto res = pages_.insert(page);
+        vassert(
+          res.second,
+          "Page already in index offset {} size {}",
+          page->offset(),
+          page->size());
+
+        sched_->submit_read(&queue_, page.get());
+        co_await waiter->ready.get_future();
+    }
+}
+
+void pager::handle_completion(page& page) noexcept {
+    if (page.test_flag(page::flags::faulting)) {
+        vassert(page.test_flag(page::flags::read), "Expected read page");
+        page.clear_flag(page::flags::faulting);
+        page.signal_waiters();
+    } else if (page.test_flag(page::flags::dirty)) {
+        vassert(page.test_flag(page::flags::write), "Expected write page");
+        if (!page.test_flag(page::flags::queued)) {
```

Comment:
the io-queue clears the queued flag before write-back happens. normally the write finishes and the completion handler is called and the flag is still cleared.

but while the write was happening, the page may have been re-queued because new data arrived and was written into the page, in which case the queued flag will be set. this is also the case here in which we don't want to incorrectly clear the dirty flag.

now that i'm thinking about it, it might be simpler to only invoke the completion handler for writes when the page wasn't re-queued.

Reply:
fixed

Reply:
Thanks for explaining, that's really helpful. I agree, it may make for easier reasoning to have the io-queue only call the completion callback when the page isn't requeued. Does this result in a meaningful behavioral change for users? Is it that writers now have to wait longer in the face of concurrent writers, before getting a page as we're appending? (I think that makes sense, if so)

Reply:
> Does this result in a meaningful behavioral change for users? Is it that writers now have to wait longer in the face of concurrent writers, before getting a page as we're appending? (I think that makes sense, if so)

For writers, there should be no change: new writes can always land into a dirty page as appends. That's true even if the page is queued for write-back.

For readers, there should be no change: dirty data in the cache is always readable.

Reply:
Ah makes sense. I saw us calling `get_page()` on the write path and assumed that that could lead us to waiting on the write path. But it makes sense that we could return a dirty page, in which case we aren't faulting (and not waiting!)

Reply:
yep. i think that there is a good argument for having a fast path for `get_page` that doesn't use `co_await` which would make it more clear.